engine/robot_logic/traversal.py

```diff
@@ -208,64 +212,98 @@ def update_state(robot_state, velocity, omega):
     """Move the robot in a roomba-like manner.
     Args:
         base_station_loc (Tuple): The (x,y) location of the base station
         time_limit (Double): How long the robot will move using roomba mode
         roomba_radius (Double): Maximum radius from the base station that the robot can move
     Returns:
-        None
+        robot_state
     """
     # to add obstacle avoidance, can either change roomba to go to point on the circle in the direction of the
     # robot direction or use raw sensor value
     # can we guarantee that when giving a velocity that the robot will move at that velocity?
-    dt = 0
+    dt = 10
+    accumulated_time = 0
     # TODO: battery_limit, time_limit, tank_capacity is full
     exit_boolean = False
-    from electrical.ultrasonic_sensor import Ultrasonic
+    if not robot_state.is_sim:
+        from electrical.ultrasonic_sensor import Ultrasonic
 
-    front_ultrasonic = Ultrasonic(0)
+        front_ultrasonic = Ultrasonic(0)
+    curr_pos = past_pos = robot_state.state
     while not exit_boolean:
         # sensor should not detect something in the robot
-        if front_ultrasonic.distance() < robot_state.front_sensor_offset:
+        if (
+            not robot_state.is_sim
+            and front_ultrasonic.distance() < robot_state.front_sensor_offset
+        ):
             robot_state.phase = Phase.FAULT
             phase_change(robot_state)
             return None
-        curr_x = robot_state.state[0]
-        curr_y = robot_state.state[1]
-        curr_head = robot_state.state[2]
-        [new_x, new_y, new_theta] = robot_state.ekf.get_predicted_state(
-            [curr_x, curr_y, curr_head],
-            [robot_state.move_dist * robot_state.time_step, 0],
-        )
+        curr_pos = robot_state.state
+        vel = calculate_dist(past_pos[:2], curr_pos[:2]) / (dt / 1000)
+        if robot_state.using_ekf:
+            [new_x, new_y, _] = robot_state.ekf.get_predicted_state(
+                curr_pos, [vel * dt / 1000, 0]
+            )
+        else:
+            [new_x, new_y, _] = curr_pos + vel * dt / 1000
         next_radius = calculate_dist(base_station_loc, (new_x, new_y))
         robot_state.goal_location = (new_x, new_y)
         # if moving will cause the robot to move through the obstacle
-        is_next_timestep_blocked = next_radius < robot_state.detect_obstacle_range
+        if not robot_state.is_sim:
+            # if obstacle detected; have to make sure move_dist < detect_obstacle_range; otherwise, robot might hit obstacle before we go back to this loop
+            is_next_timestep_blocked = (
+                front_ultrasonic.distance() < robot_state.detect_obstacle_range
+            )
+        else:
+            next_radius_2 = calculate_dist(base_station_loc, curr_pos[:2])
+            threshold = 1
+            is_next_timestep_blocked = abs(next_radius_2 - roomba_radius) < threshold
```

Comment:
the changes to is_next_timestep_blocked make sense but why don't we also want to check for the old case: next_radius < robot_state.detect_obstacle_range? don't we want to check if our future movement will result in moving through the obstacle? I know we check for next_radius > roomba_radius in the line below but I think this is a separate case right?